APM > .NET Tracer (#11200)
DataDog/documentation: content/en/tracing/setup_overview/compatibility_requirements/dotnet-core.md

</div>

## Compatibility

- The .NET Tracer supports all .NET-based languages (for example, C#, F#, Visual Basic).

- The .NET Tracer library for Datadog is open-source. For more information, see the [tracer Github repository][1].

### Supported .NET Core runtimes
The .NET Tracer supports automatic instrumentation on the following .NET Core versions. It also supports [.NET Framework][2].

| Version              | Microsoft End of Life |
| -------------------- | --------------------- |
| .NET 5               |                       |
| .NET Core 3.1        | 12/03/2022            |
| .NET Core 2.1        | 08/21/2021            |

 Additional information on .NET Core support policy can be found at [Microsoft .NET Core Lifecyle Policy][3]. 

### Supported Processor Architectures

> The .NET Tracer supports automatic instrumentation on the following architectures.

Reviewer: ```suggestion
The .NET Tracer supports automatic instrumentation on the following architectures:
```

      tag: 'GitHub'
      text: 'Examples of Custom Instrumentation'
---
<div class="alert alert-warning">
  <strong>Notes:</strong><br><ul><li>Datadog automatic instrumentation relies on the .NET CLR Profiling API. This API allows only one subscriber (for example, APM). To ensure maximum visibility, run only one APM solution in your application environment.</li><li> If you are using both automatic and custom instrumentation, it is important to keep the package versions (for example, MSI and NuGet) in sync.</li></ul>

> </div>

Reviewer 2: +1, and since this is now brought to my attention, I'd suggest moving this from the compatibility page to the setup page. That would make the most sense to me

> </div>

Author: Updated on both .NET Core & Framework and moved this out of the compatibility page to the setup page under the Installation and getting started section.

## Compatibility

- The .NET Tracer supports all .NET-based languages (for example, C#, F#, Visual Basic).

- The .NET Tracer library for Datadog is open-source. For more information, see the [tracer Github repository][1].

> ### Supported .NET Core runtimes

Author: The reason I left the Compatibility section was to provide space for the following lines: 

- The .NET Tracer supports all .NET-based languages (for example, C#, F#, Visual Basic).

- The .NET Tracer library for Datadog is open-source. For more information, see the tracer Github repository.

Otherwise they would not have a section header.

> The .NET Tracer supports instrumentation on .NET Core 2.1 and above. It also supports [.NET Framework][2].

Reviewer 3: ```suggestion
The .NET Tracer supports automatic instrumentation on the following .NET Core versions. It also supports [.NET Framework][2].
```

The .NET Tracer supports automatic instrumentation on the following .NET Core versions. It also supports [.NET Framework][2].

| Version              | Microsoft End of Life |
| -------------------- | --------------------- |
| .NET 5               |                       |
| .NET Core 3.1        | 12/03/2022            |
| .NET Core 2.1        | 08/21/2021            |

 Additional information on .NET Core support policy can be found at [Microsoft .NET Core Lifecyle Policy][3]. 

> ### Supported Processor Architectures

Author: Replaced.

> The following list only applies to automatic instrumentation:

Reviewer 3: ```suggestion
The .NET Tracer supports automatic instrumentation on the following architectures.
```

| ASP.NET Core                    | `Microsoft.AspNetCore`</br>`Microsoft.AspNetCore.App`</br>2.0+ and 3.0+                   | `AspNetCore`         |
| AWS SQS                         | `AWSSDK.SQS`  3.0+                                                                        | `AwsSqs`             |
| CosmosDb                        | `Microsoft.Azure.Cosmos.Client` 3.6.0                                                     | `CosmosDb`           |
| Elasticsearch                   | `Elasticsearch.Net` 5.3.0+                                                                | `ElasticsearchNet`   |
| HttpClient / HttpMessageHandler | `System.Net.Http` 4.0+                                                                    | `HttpMessageHandler` |
| Kafka                           | `Confluent.Kafka` 1.4+                                                                    | `Kafka`              |
| MongoDB                         | `MongoDB.Driver.Core` 2.1.0+                                                              | `MongoDb`            |

> | MSMQ                            | built-in                                                                                  | `Msmq`               |

Reviewer 3: MSMQ client it not included in .NET Core.

```suggestion
```

      tag: 'GitHub'
      text: 'Examples of Custom Instrumentation'
---
<div class="alert alert-warning">

>   <strong>Notes:</strong><br><ul><li>Datadog automatic instrumentation relies on the .NET CLR Profiling API. This API allows only one subscriber (for example, APM). To ensure maximum visibility, run only one APM solution within your application environment.</li><li> If you are using both automatic and custom instrumentation, it is important to keep the package versions (for example, MSI and NuGet) in sync.</li></ul>

Reviewer 3: ```suggestion
  <strong>Notes:</strong><br><ul><li>Datadog automatic instrumentation relies on the .NET CLR Profiling API. This API allows only one subscriber (for example, APM). To ensure maximum visibility, run only one APM solution in your application environment.</li><li> If you are using both automatic and custom instrumentation, it is important to keep the package versions (for example, MSI and NuGet) in sync.</li></ul>
```

</div>

## Compatibility

- The .NET Tracer supports all .NET-based languages (for example, C#, F#, Visual Basic).

- The .NET Tracer library for Datadog is open-source. For more information, see the [tracer Github repository][1].

### Supported .NET Core runtimes
The .NET Tracer supports instrumentation on .NET Core 2.1 and above. It also supports [.NET Framework][2].

| Version              | Vendor End of Life |
| -------------------- | ------------------ |
| .NET 5               |                    |
| .NET Core 3.1 (LTS)  | 12/03/2022         |
| .NET Core 3.0        | 03/03/2020         |

> | .NET Core 2.2        | 12/23/2019         |

Reviewer 3: This now makes it look like we support 2.2 and 3.0, which we don't.

Also, if this page is about what _Datadog_ supports, why are we adding the dates when Microsoft drops support for any runtime version so prominently? Is this helpful for users or just noise?

Author: I am happy to change the vendor to Microsoft end of life. I do think there is value here to show users that these versions have reached EOL by Microsoft.

Author: Updated vendor to mean Microsoft in https://github.com/DataDog/documentation/pull/11200/commits/77793e5a94b644a2679fb90ab651acd9a154876b 

I still see the value of having EOL from Microsoft shared with our customers. If we decide later its not useful then we can remove it.

| .NET Core 2.1 (LTS)  | 08/21/2021         |

 Additional information on .NET Core support policy can be found at [Microsoft .NET Core Lifecyle Policy][3]. 

### Supported Process Architecture

| Processor architectures                                                                                            |
| ------------------------------------------------------------------------------------------------------------------ |
| Windows x86 (`win-x86`)                                                                                            |
| Windows x64 (`win-x64`)                                                                                            |
| Linux x64 (`linux-x64`)                                                                                            |
| Alpine Linux x64 (`linux-musl-x64`)                                                                                |

> | Linux ARM64 (`linux-arm64`)<br><br>Added in version 1.27.0, automatic instrumentation only supported on .NET 5.    |

Reviewer 3: It wasn't clear before, but this entire list only applies to automatic instrumentation (manual instrumentation is processor-agnostic). Calling it out like this is misleading. I think we should add a caption to the table, and this note can be simplified to ".NET 5 only, added in version 1.27.0."

Author: Added a note and updated Linux Arm64 comment. Will be reflected in new PR.

Author: Updated and reflected in: https://github.com/DataDog/documentation/pull/11200/commits/77793e5a94b644a2679fb90ab651acd9a154876b

| MySql                           | `MySql.Data` 6.7.0+                                                                       | `AdoNet`             |
| Oracle                          | `Oracle.ManagedDataAccess` 4.122.0+                                                       | `AdoNet`             |
| PostgreSQL                      | `Npgsql` 4.0+                                                                             | `AdoNet`             |
| RabbitMQ                        | `RabbitMQ.Client` 3.6.9+                                                                  | `RabbitMQ`           |
| Redis (ServiceStack client)     | `ServiceStack.Redis` 4.0.48+                                                              | `ServiceStackRedis`  |
| Redis (StackExchange client)    | `StackExchange.Redis` 1.0.187+                                                            | `StackExchangeRedis` |
| Service Fabric Remoting         | `Microsoft.ServiceFabric.Services.Remoting` 4.0.470+                                      | `ServiceRemoting`    |

> | SQL Server                      | `System.Data` 4.0.0+, `System.Data.SqlClient` 4.0.0 +, `Microsoft.Data.SqlClient` 1.0.0+  | `AdoNet`             |

Reviewer 3: The line breaks (`</br>`) for removed here, which messes up the layout.
![image](https://user-images.githubusercontent.com/1851278/127042963-453a4c9d-b321-42af-99cf-cc78e4381732.png)
```suggestion
| SQL Server                      | `System.Data` 4.0.0+</br>`System.Data.SqlClient` 4.0.0+</br>`Microsoft.Data.SqlClient` 1.0.0+  | `AdoNet`             |
```

      tag: 'GitHub'
      text: 'Examples of Custom Instrumentation'
---
<div class="alert alert-warning">
  <strong>Notes:</strong><br><ul><li>Datadog automatic instrumentation relies on the .NET CLR Profiling API. This API allows only one subscriber (for example, APM). To ensure maximum visibility, run only one APM solution within your application environment.</li><li> If you are using both automatic and custom instrumentation, it is important to keep the package versions (for example, MSI and NuGet) in sync.</li></ul>
</div>

## Compatibility
Datadog .NET application performance monitoring supports a robust set of features to provide critical information about .NET applications & services: 

- The .NET Tracer supports all .NET-based languages (for example, C#, F#, Visual Basic).

> - Monitor [runtime metrics][1] to help troubleshoot bugs and detect resource inefficiencies in your application.

Reviewer 2: My preference is to remove these new bullets and have the compatibility page be as minimal as possible, that solely answer the question of "Will this run on my machines?"

Author: Interesting, I do agree on the point of making it as clear as possible. However, having the capabilities section include the Runtime Metics & logs injection, which are  unique differentiators to our competitors does also serve a purpose. My goal with this is to help enable global customers, sales teams in different regions, and support engineers with all the features the .NET APM library has to offer.

Reviewer 4: I guess this depends on whether we think this page should be focused on sales/new prospects, or whether it's devs looking to implement it. My assumption is the latter, and so I'm inclined to agree with Zach, especially under the "compatibility" title

Author: Thanks team, I agree. I will add a new section for Capabilities above the Compatibility section. I will draft up a PR now - again, thank you!

Author: Update: the changes have been added to the following [PR](https://github.com/DataDog/documentation/pull/11200/commits/3cbbb246f859f9a4102decfd6272a0a510cd777a).  I moved the Capabilities section to the main docs and left the Compatibility minimal. 

Marking this as resolved.

- Ability to correlate [traces & application logs][2].

- The .NET Tracer library for Datadog is open-source. For more information, see the [tracer Github repository][3].

> ### .NET runtimes

Reviewer 2: Maybe "Supported .NET Runtimes"?

Author: Possibly, I thought it would of been inferred.

Author: Changed have been updated [here](https://github.com/DataDog/documentation/pull/11200/commits/d4260a023194b2e0791277b5c9242b0903fff21e). 

Marking as resolved.

Datadog .NET Tracer supports instrumentation for .NET Core & [NET Framework][4]. 

| Version                         | Vendor End of Life |
| ------------------------------- | ------------------ |
| .NET 5                          |                    |

> | .NET Core 3.1                   | 12/03/2022         |

Reviewer 4: 3.1 is _also_ an LTS release

Author: Ah, good catch - thank you! I will update that later.

Author: Changes have been added [here](https://github.com/DataDog/documentation/pull/11200/commits/d4260a023194b2e0791277b5c9242b0903fff21e)

Marking as resolved

| .NET Core 3.0                   | 03/03/2020         |
| .NET Core 2.2                   | 12/23/2019         |
| .NET Core 2.1 (LTS)             | 08/21/2021         |

 Additional information on .NET Core support policy can be found at [Microsoft .NET Core Lifecyle Policy][5]. 

### Supported Process Architecture

> | Processor architectures                                                                                            |

Reviewer 2: I think this could be improved by having two columns, something like this:

```
| Processor architecture           | First Supported Runtime |
| ------------------------------- | ------------------ |
| Windows x86                         | .NET Core 2.1      |
| Windows x64                         | .NET Core 2.1      |
| Linux x64                               | .NET Core 2.1      |
| Alpine Linux x64                    | .NET Core 2.1      |
| Linux ARM64                         | .NET 5      |
```

Author: I thought about this but I didn't want it to be preserved as a specific version only works for a specific process architecture.

Reviewer 2: Yeah that makes sense, resolving

      text: 'Examples of Custom Instrumentation'
---

- The .NET Tracer supports all .NET-based languages (for example, C#, F#, Visual Basic).

> - The .NET Tracer library for Datadog is open-source. For more information, see the [tracer Github repository][1].

Reviewer 5: ```suggestion
- The .NET Tracer library for Datadog is open source. For more information, see the [.NET Tracer repository][1].
```

## Supported .NET Core runtimes
The .NET Tracer supports automatic instrumentation on the following .NET Core versions. It also supports [.NET Framework][2].

| Version              | Microsoft End of Life |
| -------------------- | --------------------- |
| .NET 5               |                       |
| .NET Core 3.1        | 12/03/2022            |
| .NET Core 2.1        | 08/21/2021            |

>  Additional information on .NET Core support policy can be found at [Microsoft .NET Core Lifecyle Policy][3]. 

Reviewer 5: ```suggestion
 Additional information on .NET Core support policy can be found within [Microsoft's .NET Core Lifecycle Policy][3]. 
```
